programs/drift/src/state/lp_pool.rs

```diff
@@ -0,0 +1,1462 @@
+    }
+
+    fn increase_balance(&mut self, delta: u128) -> DriftResult {
+        self.scaled_balance = self.scaled_balance.safe_add(delta)?;
+        Ok(())
+    }
+
+    fn decrease_balance(&mut self, delta: u128) -> DriftResult {
+        self.scaled_balance = self.scaled_balance.safe_sub(delta)?;
+        Ok(())
+    }
+
+    fn update_balance_type(&mut self, balance_type: SpotBalanceType) -> DriftResult {
+        self.balance_type = balance_type;
+        Ok(())
+    }
+}
+
+impl BLPosition {
+    pub fn get_token_amount(&self, spot_market: &SpotMarket) -> DriftResult<u128> {
+        get_token_amount(self.scaled_balance, spot_market, &self.balance_type)
+    }
+}
+
+#[account(zero_copy(unsafe))]
+#[derive(Default, Debug, BorshDeserialize, BorshSerialize)]
+#[repr(C)]
+pub struct Constituent {
+    /// address of the constituent
+    pub pubkey: Pubkey,
+    pub mint: Pubkey,
+    pub lp_pool: Pubkey,
+    pub token_vault: Pubkey,
+
+    /// total fees received by the constituent. Positive = fees received, Negative = fees paid
+    pub total_swap_fees: i128,
+
+    /// spot borrow-lend balance for constituent
+    pub spot_balance: BLPosition, // should be in constituent base asset
+
+    /// max deviation from target_weight allowed for the constituent
+    /// precision: PERCENTAGE_PRECISION
+    pub max_weight_deviation: i64,
+    /// min fee charged on swaps to/from this constituent
+    /// precision: PERCENTAGE_PRECISION
+    pub swap_fee_min: i64,
+    /// max fee charged on swaps to/from this constituent
+    /// precision: PERCENTAGE_PRECISION
+    pub swap_fee_max: i64,
+
+    /// ata token balance in token precision
+    pub token_balance: u64,
+
+    pub last_oracle_price: i64,
+    pub last_oracle_slot: u64,
+
+    pub oracle_staleness_threshold: u64,
+
+    pub flash_loan_initial_token_amount: u64,
+    /// Every swap to/from this constituent has a monotonically increasing id. This is the next id to use
+    pub next_swap_id: u64,
+
+    /// percentable of derivatve weight to go to this specific derivative PERCENTAGE_PRECISION. Zero if no derivative weight
+    pub derivative_weight: u64,
+
+    pub volatility: u64, // volatility in PERCENTAGE_PRECISION 1=1%
+
+    pub constituent_derivative_index: i16, // -1 if a parent index
+
+    pub spot_market_index: u16,
+    pub constituent_index: u16,
+
+    pub decimals: u8,
+    pub bump: u8,
+
+    // Fee params
+    pub gamma_inventory: u8,
+    pub gamma_execution: u8,
+    pub xi: u8,
+    pub _padding: [u8; 5],
+}
+
+impl Size for Constituent {
+    const SIZE: usize = 288;
+}
+
+impl Constituent {
+    /// Returns the full balance of the Constituent, the total of the amount in Constituent's token
+    /// account and in Drift Borrow-Lend.
+    pub fn get_full_balance(&self, spot_market: &SpotMarket) -> DriftResult<i128> {
+        match self.spot_balance.balance_type() {
+            SpotBalanceType::Deposit => self.token_balance.cast::<i128>()?.safe_add(
```

Comment:
can just implement get_signed_token_amount on spot_balance and that will give you a i128